### ISSUE TITLE
Modal should use icon tooltip

### ISSUE THREAD

**Issue by elizabethsjudd** [carbon-design-system/carbon#3630]
## Environment

> Operating system

Mac OS X

> Browser

Chrome, Safari, FF

> Automated testing tool and ruleset

AAT July 2019 ruleset
WCAG 2.1

> Assistive technology used to verify

keyboard, mouse

## Detailed description

The closed button on modals should use the icon tooltip to provide a textual label for sighted users

> What version of the Carbon Design System are you using?

v10.4.1

> What did you expect to happen?

There is a textual label for sighted users on icon only buttons

> What happened instead?

the close button only has a label for screen reader users

> What WCAG 2.1 checkpoint does the issue violate?

## Steps to reproduce the issue

1. Interact (hover/focus) the modal close button

> Please create a reduced test case in CodeSandbox
>
> - Style and vanilla JS:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/components/examples/codesandbox
> - React:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/react/examples/codesandbox

## Additional information

@carbon-design-system/developers you can assign this issue to me and I can take a look at it in the next couple of weeks

**Comment by vpicone** [comment]
@elizabethsjudd Thanks for contributing! For these a11y issues do you mind linking to the specific rule being broken? That way we're not guessing at the success criteria. This is my best guess at the rule you're trying to address here:

**WCAG 2.1 rule `4.1.2 Name, Role, Value (Level A)`:**
> Success Criterion: For all user interface components (including but not limited to: form elements, links and components generated by scripts), the name and role can be programmatically determined; states, properties, and values that can be set by the user can be programmatically set; and notification of changes to these items is available to user agents, including assistive technologies.

Under `Sufficient Techniques`, you'll find the first sufficient technique is our [aria-label](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA14) approach. That approach addresses requirements for **sighted users**:

> For sighted users, the context and visual appearance of an element can provide sufficient cues to determine the purpose. **An example is the 'X' often used in the top right corner of pop-up divs (light boxes) to indicate the control for closing the div.**

Conveniently, the first [example they present](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA14#examples) is a modal that uses an X and aria-label (our approach in both react and Vanilla)

Let me know if this wasn't the rule you were referring to!

**Comment by elizabethsjudd** [comment]
This issue is more about consistency within the system as well as a11y enhancements versus just a full violation.  I contributed changes to Icon Tooltip to specifically not use the `aria-label` attribute because of updates to AAT that came out with WCAG 2.1 where there are violations for buttons that use the `aria-label` attribute. This enhancement is for cognitive disabilities to provide textual descriptions for sighted users.  A large focus in a11y updates are around considerations for cognitive disabilities as the use of icons and placeholder text has become increasingly popular in UI designs.

**Comment by vpicone** [comment]
> updates to AAT that came out with WCAG 2.1 where there are violations for buttons that use the aria-label attribute

This seems at odds with the close-modal-button example from the WCAG guidelines I linked above. Do you have a resource where I could learn more about this?

The Cognitive and Learning Disability task published a **draft** document last week that recommends [tooltips for icons and jargon](https://w3c.github.io/coga/techniques/index.html#include-short-tooltips-on-all-icons-jargon), but I'm wary of implementing their recommendation at this point for a few reasons:

1. > The group does not expect this document to become a W3C Recommendation.

2. > This is a draft document and may be updated, replaced or obsoleted by other documents at any time. 
	Removing attributes or changing the DOM structure of components is something we can't take lightly as they are technically breaking changes.

3. The guideline suggests that to be helpful, tooltips should [be optional]([src](https://w3c.github.io/coga/techniques/index.html#include-short-tooltips-on-all-icons-jargon)):
	> they could add clutter and even confuse if they are read out by a screen reader or text to speech support

**Comment by elizabethsjudd** [comment]
WCAG are requirements for what are the minimum for a11y and have found that "just passing the tests" doesn't always result in the best user experience. I'd like to get feedback on this issue from the IBMa team. @snidersd thoughts on this issue and #3631 (same issue but for Notification component)?

Also all of these changes can be implemented easily in a backwards compatible way so even though they are HTML changes it doesn't mean that it's going to break anyone's code using carbon but would be able to provide a clearer experience for the users that have the newer markup.  I also feel like this is an improvement for internationalization as it allows a translated textual description for sighted users as well.

**Comment by snidersd** [comment]
@elizabethsjudd It looks to me like this is an issue specific to the vanilla implementation. When I look at the Modal and Notification components in React the tooltip is present on hover. 
<img width="877" alt="Screen Shot 2019-08-02 at 9 31 57 AM" src="https://user-images.githubusercontent.com/21676914/62378323-2cf40200-b50a-11e9-8f5e-02e55518eab9.png">
This same behavior should be applied to the vanilla component.

**Comment by elizabethsjudd** [comment]
@snidersd it looks like the react version is using the `title` (works for mouse users) and `aria-label` (works for screen reader users) attributes. I'm proposing we use the `icon tooltip` component so that it shows up for sighted keyboard users as well.

**Comment by vpicone** [comment]
> even though they are HTML changes it doesn't mean that it's going to break anyone's code using carbon

Unfortunately this isn't the case, people use aria attributes for styling all the time so we need to be cautious about removing/changing them and breaking their styles.

My request is that you not list `WCAG 2.1` as the source for a change if they don't have rules that pertain to it. I don't think we should be making changes that aren't backed by _some_ resource. If we don't, we run the risk of mis-applying or over-applying a11y hints.

**Comment by elizabethsjudd** [comment]
@vpicone The icon tooltip update was tied to a specific WCAG 2.1 change with the use of `aria-label` that was causing an AAT 2.4.1 violation. The update to no longer throw this violation in AAT has already be released and it 100% is backwards compatible otherwise I wouldn't have been able to merge it.  We did thorough testing to make sure that the style updates worked for both HTML structures.

The Icon Tooltip was approved as the preferred way to create icon-only buttons and all i'm asking is that Carbon be consistent.  There are multiple icon-only buttons throughout Carbon and they should all use the same format to keep a consistent user experience.  I want this applied holistically across Carbon but have been going component by component (it's already been added to overflow menu in 10.4.1) to make the work more manageable.

**Comment by vpicone** [comment]
@elizabethsjudd I looked and wasn't able to find the WCAG 2.1 change you're referring to, if you could point it out to me that'd be awesome!

> We did thorough testing to make sure that the style updates worked for both HTML structures.

It's not just Carbon's styles, it's styles from people using carbon. They're free to target whatever attributes they wish.

I'm not familiar with the OverflowMenu change you got through. Does that mean right now we're split in our approach in this behavior across different components? That inconsistency is [explicitly against guidance from the cognitive working group](https://w3c.github.io/coga/techniques/index.html#use-consistent-styles). Applying this holistically would mean instead applying it to all the components simultaneously.

**Comment by elizabethsjudd** [comment]
I wasn't told that I needed to do it all in one PR but if I need to I can, I just need to know a list of components this affects (i'm aware of Modal and Notification which is why I logged those issues).  Carbon users don't need to update to the new structure so they can still target aria-label if they want to until the old syntax is depreciated and removed. This is simply an enhancement to support cognitive and sighted keyboard users.

There is a whole conversation between Josh and I in this PR about WCAG and why I made the update to the tooltip: https://github.com/carbon-design-system/carbon/pull/3115

**Comment by vpicone** [comment]
I've read through that PR as well as the new guidelines and am still not clear on [which new success criteria](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-21/) you're referring to. 

Is the error getting thrown here as well? Again, I think it's odd that the WCAG rules are causing a violation when our solution uses their first success technique (the close modal link above) verbatim.

If a rule is coming from advice you got from a webinar we need to know that rather than just saying WCAG 2.1 assuming it's true.

**Comment by elizabethsjudd** [comment]
That is the violation: buttons are an input.  And meeting "sufficient" criteria doesn't mean that it's the best experience. There are plenty of examples on their site that are not the best experience or work for all use-cases especially around modals.  Sharon even said that displaying the tooltip with title was better than what was currently happening in Vanilla modals... we can give sighted keyboard users the better experience as well by using the icon tooltip over the title attribute.  Don't we want all users to have the best experience we can give them?

**Comment by vpicone** [comment]
@elizabethsjudd some keyboard users might find the tooltip distracting. I think this discounts their experience.

> That is the violation: buttons are an input.

Buttons (and even inputs with type of button) and inputs are actually very different. Buttons can receive labels from their value, but also have different labeling recommendations per [success criteria 3.3.2](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html).

I totally understand going beyond what is required, but as we gear up for implementing 2.1 requirements we need to distinguish between what actually comes from the WCAG recommendation and what is subjective opinion/draft material. 

Do you have a screen shot of it failing here? I’m not able to reproduce.

**Comment by elizabethsjudd** [comment]
@vpicone @snidersd using `aria-label` and `title` together can actually cause duplicate content for screen reader users in the way that it is being used in Carbon React modals.  Both JAWS 15 and VO on iOS will read both attributes' content to the user.

By using the new Icon Tooltip we are:
1. Ensuring there is no duplicate content in the DOM (the text exists within the button only)
2. Icon-only buttons are implemented consistently across all components and the experience isn't different depending on the component.
3. We treat keyboard users the same as mouse users so that all sighted users have the same textual information available no matter how they are interacting with the component

Issue thrown by AAT
<img width="1552" alt="Screen Shot 2019-08-05 at 11 09 11 AM" src="https://user-images.githubusercontent.com/1275779/62474720-8447db80-b771-11e9-9877-a1738e8eeea9.png">

**Comment by vpicone** [comment]
@elizabethsjudd that says potential violation, so is this not an actual violation? Is there any guidelines for determining whether it’s a real violation or not?

For 2 that means we would implement this in a single PR to prevent a different experience between components

For 3: it would actually be a different experience. Mouse users have to chose to hover a component to get the feedback. Keyboard users can have the button announced to them but this would force the tooltip without them asking for additional context.

**Comment by elizabethsjudd** [comment]
@vpicone Potential violation means you need to manually test it.  For mouse and screen reader users they have textual labels available. Keyboard users do not so it's an actual violation.  

WRT 2, I'm fine implementing this change across all components as long as I have a list of components that need updated.  Right now I know of modal and notification, if there are more let me know and I'll make the updates.  

WRT3, just like a mouse user chooses to hover over the button, keyboard users chose to focus it.  They are interacting with the element before the text is displayed.  Hover and focus are typically considered the same interaction/intention but for different input devices.

**Comment by elizabethsjudd** [comment]
@dakahn If you're okay with me implementing this, can this issue be assigned to me so I can work on it in the next week.  I'll combine notifications and modals in to a single PR if preferred (and any other component that also needs this update if you can let me know what they are).

**Comment by asudoh** [comment]
Jumping in - I tend to agree @snidersd on this that we should go with the solution of attributes. Going further seems to go beyond our a11y requirements IIUC.

**Comment by elizabethsjudd** [comment]
From IBMa's user testing talk today! A key pain point was lack of text labels on icons for ALL users. Using the title attribute does not provide this for keyboard users.

<img width="1276" alt="Screen Shot 2019-08-14 at 9 42 15 AM" src="https://user-images.githubusercontent.com/1275779/63025761-e1c9df80-be77-11e9-82d1-c2787746e454.png">

**Comment by snidersd** [comment]
If I am understanding this issue correctly, the problem is related to the change in requirements from WCAG 2.0, which is the current IBM requirement versus WCAG 2.1 that goes into effect January 2020. 
The ARIA Authoring Practices (APG) currently provide guidance for WCAG 2.0 only. At this point, the APG has not been updated with WCAG 2.1 examples.
The React Tooltip components have issues for both WCAG 2.0 & WCAG 2.1 that need to be considered (if they haven’t been already). 
By January 2020 all tooltip issues will need to be fixed to meet the new [WCAG 2.1 SC 1.4.13: Content on Hover or Focus](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html). 
The other issues related to the tooltip that need to be addressed are #2736 #3615 and #3616. 
Note: #3615 & #3616 were the direct result of the Low Version work discussed in the presentation this morning.

**Comment by elizabethsjudd** [comment]
@snidersd This issue is we should be using the icon tooltip component on the close button (which is an icon-only button) for modals (and notifications).

Current status:
* Vanilla doesn't use the title attribute so no sighted users can access the label
* React uses `title` attribute which creates duplicate content for some screen reader users and also does not support sighted keyboard users

Proposed change:
* **ALL** icon only buttons use the Icon Tooltip component so there is (1) consistent UX for all icon only buttons, (2) supports keyboard users and screen reader users don't have duplicated content, (3) future updates to Icon Tooltip for issues discussed in #3615 and #3616 would automatically be added across the board.

I understand that this may not be part of WCAG 2.0 which is the current requirement but as of January WCAG 2.1 is going to be the requirement and we should implement this now so we are not cramming in December.

**Comment by dakahn** [comment]
Closing this issue since as is it's not something we're going to be working on in the foreseeable future. We _do need_ gather requirements for 2.1 updates, but even after reading all these comments I'm unclear if this is an actual requirement -- UX consistency or whatever else aside. 

Does the icon button in a modal require tooltip labeling? 

If not it sounds like a cool enhancement for sighted users.

**Comment by elizabethsjudd** [comment]
@dakahn this is something that I offered to work on and was not expecting the core team to do. If it's agreed that it's a good enhancement for sighted users I don't understand why I can't do the work.

**Comment by dakahn** [comment]
@elizabethsjudd [I've opened an enhancement ticket describing this as a value add for sighted users](https://github.com/carbon-design-system/carbon/issues/4173). Please feel free comment with any research or background information. I think we can have a more meaningful conversation that way.

Enhancements are triaged and considered differently than bug tickets. I speak for everyone on the team when I say that we're overjoyed to have anyone contribute anything meaningful.
